File: python-client/calculator_client.py
import grpc
import calculator_pb2
import calculator_pb2_grpc
import sys
import os

def run_calculator_operations(server_host='localhost:50051'):
    # Create a gRPC channel
    with grpc.insecure_channel(server_host) as channel:
        # Create a stub (client)
        stub = calculator_pb2_grpc.CalculatorStub(channel)
        
        print("=== Python Calculator Client ===")
        print(f"Connected to server at {server_host}")
        
        # Test different operations
        test_cases = [
            (10.5, 5.2, "addition"),
            (15.8, 3.3, "subtraction"), 
            (7.5, 4.0, "multiplication"),
            (20.0, 4.0, "division"),
            (10.0, 0.0, "division")  # Test division by zero
        ]
        
        for a, b, operation in test_cases:
            try:
                request = calculator_pb2.CalculationRequest(a=a, b=b)
                
                if operation == "addition":
                    response = stub.Add(request)
                elif operation == "subtraction":
                    response = stub.Subtract(request)
                elif operation == "multiplication":
                    response = stub.Multiply(request)
                elif operation == "division":
                    response = stub.Divide(request)
                
                print(f"\n{operation.upper()}:")
                print(f"Request: a={a}, b={b}")
                print(f"Response: {response.message}")
                print(f"Result: {response.result}")
                
            except grpc.RpcError as e:
                print(f"RPC failed for {operation}: {e}")

def interactive_mode(server_host='localhost:50051'):
    with grpc.insecure_channel(server_host) as channel:
        stub = calculator_pb2_grpc.CalculatorStub(channel)
        
        print("\n=== Interactive Calculator Mode ===")
        print("Commands: add, subtract, multiply, divide, quit")
        
        while True:
            operation = input("\nEnter operation (or 'quit'): ").strip().lower()
            
            if operation == 'quit':
                break
                
            if operation not in ['add', 'subtract', 'multiply', 'divide']:
                print("Invalid operation. Use: add, subtract, multiply, divide, quit")
                continue
                
            try:
                a = float(input("Enter first number: "))
                b = float(input("Enter second number: "))
                
                request = calculator_pb2.CalculationRequest(a=a, b=b)
                
                if operation == 'add':
                    response = stub.Add(request)
                elif operation == 'subtract':
                    response = stub.Subtract(request)
                elif operation == 'multiply':
                    response = stub.Multiply(request)
                elif operation == 'divide':
                    response = stub.Divide(request)
                
                print(f"Result: {response.message}")
                
            except ValueError:
                print("Please enter valid numbers")
            except grpc.RpcError as e:
                print(f"RPC failed: {e}")

if __name__ == '__main__':
    server_address = os.getenv('GRPC_SERVER', 'localhost:50051')
    
    if len(sys.argv) > 1 and sys.argv[1] == '--interactive':
        interactive_mode(server_address)
    else:
        run_calculator_operations(server_address)

File: nodejs-client/calculator-client.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const path = require('path');

// Load the protobuf
const PROTO_PATH = './proto/calculator.proto';

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true
});

const calculator = grpc.loadPackageDefinition(packageDefinition).calculator;

// Get server address from environment variable or use default
const SERVER_ADDRESS = process.env.GRPC_SERVER || 'localhost:50051';

function runCalculatorOperations() {
    // Create a client
    const client = new calculator.Calculator(SERVER_ADDRESS, grpc.credentials.createInsecure());
    
    console.log('=== Node.js Calculator Client ===');
    console.log(`Connected to server at ${SERVER_ADDRESS}`);
    
    // Test cases
    const testCases = [
        { a: 15.5, b: 7.2, operation: 'Add' },
        { a: 22.8, b: 5.3, operation: 'Subtract' },
        { a: 6.5, b: 3.0, operation: 'Multiply' },
        { a: 25.0, b: 5.0, operation: 'Divide' },
        { a: 12.0, b: 0.0, operation: 'Divide' }  // Test division by zero
    ];
    
    let completed = 0;
    const total = testCases.length;
    
    testCases.forEach((testCase, index) => {
        const request = { a: testCase.a, b: testCase.b };
        
        // Call the appropriate method
        const methodName = testCase.operation.toLowerCase();
        
        client[methodName](request, (error, response) => {
            if (error) {
                console.error(`\nError in ${testCase.operation}:`, error);
            } else {
                console.log(`\n${testCase.operation.toUpperCase()}:`);
                console.log(`Request: a=${testCase.a}, b=${testCase.b}`);
                console.log(`Response: operation=${response.operation}, result=${response.result}`);
            }
            
            completed++;
            if (completed === total) {
                client.close();
                console.log('\n=== All operations completed ===');
            }
        });
    });
}

function interactiveMode() {
    const client = new calculator.Calculator(SERVER_ADDRESS, grpc.credentials.createInsecure());
    const readline = require('readline');
    
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    console.log('\n=== Interactive Calculator Mode ===');
    console.log('Commands: add, subtract, multiply, divide, quit');
    
    function promptUser() {
        rl.question('\nEnter operation (or \'quit\'): ', (operation) => {
            operation = operation.trim().toLowerCase();
            
            if (operation === 'quit') {
                client.close();
                rl.close();
                return;
            }
            
            if (!['add', 'subtract', 'multiply', 'divide'].includes(operation)) {
                console.log('Invalid operation. Use: add, subtract, multiply, divide, quit');
                promptUser();
                return;
            }
            
            rl.question('Enter first number: ', (aStr) => {
                rl.question('Enter second number: ', (bStr) => {
                    const a = parseFloat(aStr);
                    const b = parseFloat(bStr);
                    
                    if (isNaN(a) || isNaN(b)) {
                        console.log('Please enter valid numbers');
                        promptUser();
                        return;
                    }
                    
                    const request = { a: a, b: b };
                    
                    client[operation](request, (error, response) => {
                        if (error) {
                            console.error('RPC failed:', error);
                        } else {
                            console.log(`Result: ${response.operation} = ${response.result}`);
                        }
                        promptUser();
                    });
                });
            });
        });
    }
    
    promptUser();
}

// Check command line arguments
if (process.argv.includes('--interactive')) {
    interactiveMode();
} else {
    runCalculatorOperations();
}